src/coreComponents/constitutive/contact/BartonBandis.hpp

```diff
@@ -0,0 +1,149 @@
+
+  /**
+   * @brief Evaluate the effective aperture, and its derivative wrt aperture
+   * @param[in] aperture the model aperture/gap
+   * @param[out] dHydraulicAperture_dAperture the derivative of the effective aperture wrt aperture
+   * @return The hydraulic aperture that is always > 0
+   */
+  GEOS_HOST_DEVICE
+  real64 computeHydraulicAperture( real64 const aperture,
+                                   real64 const normalTraction,
+                                   real64 & dHydraulicAperture_aperture,
+                                   real64 & dHydraulicAperture_dNormalStress ) const;
+
+private:
+  real64 m_referenceNormalStress;
+  
+  real64 m_aperture0;                                    
+};
+
+
+/**
+ * @class BartonBandis
+ *
+ * This class serves as the interface for implementing contact enforcement constitutive relations.
+ * This does not include the actual enforcement algorithm, but only the constitutive relations that
+ * govern the behavior of the contact. So things like penalty, or friction, or kinematic constraint.
+ */
+class BartonBandis : public HydraulicApertureBase
+{
+public:
+
+  /**
+   * @brief The standard data repository constructor
+   * @param name The name of the relation in the data repository
+   * @param parent The name of the parent Group that holds this relation object.
+   */
+  BartonBandis( string const & name,
+                Group * const parent );
+
+  /**
+   * @brief default destructor
+   */
+  virtual ~BartonBandis() override;
+
+
+  /// Type of kernel wrapper for in-kernel update
+  using KernelWrapper = BartonBandisUpdates;
+
+  /**
+   * @brief Create an update kernel wrapper.
+   * @return the wrapper
+   */
+  KernelWrapper createKernelWrapper() const;
+
+private:
+  
+   struct viewKeyStruct : public ConstitutiveBase::viewKeyStruct
+  {
+    /// string/key for aperture tolerance
+    static constexpr char const * apertureZeroString() { return "apertureZero"; }
+
+    /// string/key for aperture table name
+    static constexpr char const * referenceNormalStressString() { return "referenceNormalStress"; }
+  };
+
+  real64 m_referenceNormalStress;
+
+  real64 m_aperture0;                                
+
+};
+
+GEOS_HOST_DEVICE
+GEOS_FORCE_INLINE
+real64 BartonBandisUpdates::computeHydraulicAperture( real64 const aperture,
+                                                      real64 const normalTraction,
+                                                      real64 & dHydraulicAperture_aperture,
+                                                      real64 & dHydraulicAperture_dNormalStress ) const
+{
+  real64 const hydraulicAperture = ( aperture >= 0.0 ) ? (aperture + m_aperture0) : m_aperture0 / ( 1 + 9*normalTraction/m_referenceNormalStress );
+  dHydraulicAperture_dNormalStress = -hydraulicAperture / ( 1 + 9*normalTraction/m_referenceNormalStress ) * 9/m_referenceNormalStress;
+  dHydraulicAperture_aperture = 0.0;
```

Comment:
that's an excellent question to which I have no answer.... As you said it's not differentiable at that point so, by definition we can't compute the derivative.

Reply:
If we use penalty contact, `dHydraulicAperture_aperture` should be nonzero when `aperture < 0` but calculated by the penalty stiffness and `dHydraulicAperture_dNormalStress` I think. It may depend on if we want to use penalty or lagrangian or both for BartonBandis.